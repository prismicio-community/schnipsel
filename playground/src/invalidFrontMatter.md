---
name: "Hello World Invalid"
description: "Says hello to the world invalid"
scopes: false
prefix: 42
---

```typescript
export const hello = (str: string): void => {
	console.info("Hello %s", str);
};
```
