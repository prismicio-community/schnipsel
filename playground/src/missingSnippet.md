---
name: "Hello World Missing"
description: "Says hello to the world missing"
scope: ["typescript"]
prefix: "helloWorldMissing"
---
