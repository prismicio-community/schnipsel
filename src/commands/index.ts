export { _default } from "./default";
export { init } from "./init";
export { help } from "./help";
