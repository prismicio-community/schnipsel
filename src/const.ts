import {
	name as pkgName,
	description as pkgDescription,
	version as pkgVersion,
} from "../package.json";

export const NAME = pkgName;
export const PACKAGE = pkgName;
export const DESCRIPTION = pkgDescription;
export const VERSION = pkgVersion;
export const CONFIG_FILE = `${NAME}.config.ts`;
