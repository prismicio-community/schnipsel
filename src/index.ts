export { run, runWatch } from "./process";

export { ConfigFile } from "./ConfigFile";
export { Input } from "./Input";

export type { Config, SnippetObject } from "./types";
