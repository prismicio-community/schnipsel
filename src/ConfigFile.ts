import chalk from "chalk";
import { loadConfig } from "c12";

import { JSONFile, debug, AnyFile } from "./lib";
import { CONFIG_FILE, NAME } from "./const";
import { SchnipselConfig } from "./types";
import { Renderers } from "./renderers";

export class ConfigFile extends AnyFile<SchnipselConfig> {
	constructor(cwd = process.cwd()) {
		super(CONFIG_FILE, cwd);
	}

	async init(): Promise<void> {
		if (this.exists()) {
			throw new Error(
				`${chalk.cyan(this.name)} already exists in ${chalk.cyan(this.cwd)}`,
			);
		}

		const { indent } = await new JSONFile("package.json", this.cwd).readMeta();

		await this._write(
			`import { defineSchnipselConfig } from "schnipsel";

export default defineSchnipselConfig({
	input: {
		directory: "src",
	},

	// Update or remove renderers according to desired result...
	renderers: [
		{
			name: "vscode",
			options: {
				outputDirectory: "vscode/snippets",
				packageJSON: "vscode/package.json",
			},
		},
		{
			name: "sublime",
			options: {
				outputDirectory: "sublime/snippets",
			},
		},
		{
			name: "intellij",
			options: {
				outputDirectory: "intellij/src/main/resources/liveTemplates",
				pluginXML: "intellij/src/main/resources/META-INF/plugin.xml",
			},
		},
	],
});
`.replace(/\t/g, indent),
		);
	}

	async read(): Promise<SchnipselConfig> {
		debug("Reading config file...");

		if (!this.exists()) {
			throw new Error(
				`${chalk.cyan(this.name)} not found in ${chalk.cyan(
					this.cwd,
				)}, create one with ${chalk.cyan(`${NAME} init`)}`,
			);
		}

		const { config } = await loadConfig({
			name: "schnipsel",
			configFile: "schnipsel.config",
			rcFile: false,
			globalRc: false,
			cwd: this.cwd,
		});

		this.validate(config);

		debug("Config file read and validated!");

		return config;
	}

	async write(_config: SchnipselConfig): Promise<void> {
		throw new Error("`ConfigFile.write()` is not supported");
	}

	validate(
		maybeConfig: Record<string, unknown>,
	): asserts maybeConfig is SchnipselConfig {
		const errors: string[] = [];

		// Validate `input`
		if (!("input" in maybeConfig)) {
			errors.push(`${chalk.cyan("input")} is missing`);
		} else if (!maybeConfig.input || typeof maybeConfig.input !== "object") {
			errors.push(
				`${chalk.cyan("input")} should be of type ${chalk.cyan("object")}`,
			);
		} else {
			const input = maybeConfig.input as Record<string, unknown>;

			// Validate `input.directory`
			if (!("directory" in input)) {
				errors.push(`${chalk.cyan("input.directory")} is missing`);
			} else if (typeof input.directory !== "string") {
				errors.push(
					`${chalk.cyan("input.directory")} should be of type ${chalk.cyan(
						"string",
					)}`,
				);
			}

			// Validate `input.glob`
			if (
				"glob" in input &&
				typeof input.glob !== "string" &&
				!(
					Array.isArray(input.glob) &&
					input.glob.length > 0 &&
					input.glob.every((g) => typeof g === "string")
				)
			) {
				errors.push(
					`${chalk.cyan("input.glob")} should be of type ${chalk.cyan(
						"string | string[]",
					)}`,
				);
			}
		}

		if (!("renderers" in maybeConfig)) {
			errors.push(`${chalk.cyan("renderers")} is missing`);
		} else if (!Array.isArray(maybeConfig.renderers)) {
			errors.push(
				`${chalk.cyan("renderers")} should be of type ${chalk.cyan("array")}`,
			);
		} else if (maybeConfig.renderers.length === 0) {
			errors.push(`${chalk.cyan("renderers")} should not be empty`);
		} else if (
			maybeConfig.renderers.some(
				(renderer) => !renderer || typeof renderer !== "object",
			)
		) {
			errors.push(
				`All ${chalk.cyan("renderers[n]")} should be of type ${chalk.cyan(
					"object",
				)}`,
			);
		} else {
			if (maybeConfig.renderers.some((renderer) => !("name" in renderer))) {
				errors.push(`Some ${chalk.cyan("renderers[n].name")} are missing`);
			} else if (
				maybeConfig.renderers.some(
					(renderer) => typeof renderer.name !== "string",
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].name",
					)} should be of type ${chalk.cyan("string")}`,
				);
			} else if (
				maybeConfig.renderers.some(
					(renderer) => !Object.keys(Renderers).includes(renderer.name),
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].name",
					)} should be of type ${chalk.cyan(
						Object.keys(Renderers)
							.map((name) => `"${name}"`)
							.join(" | "),
					)}`,
				);
			}

			if (maybeConfig.renderers.some((renderer) => !("options" in renderer))) {
				errors.push(`Some ${chalk.cyan("renderers[n].options")} are missing`);
			} else if (
				maybeConfig.renderers.some(
					(renderer) =>
						!renderer.options || typeof renderer.options !== "object",
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].options",
					)} should be of type ${chalk.cyan("object")}`,
				);
			} else {
				if (
					maybeConfig.renderers.some(
						(renderer) => !("outputDirectory" in renderer.options),
					)
				) {
					errors.push(
						`Some ${chalk.cyan(
							"renderers[n].options.outputDirectory",
						)} are missing`,
					);
				} else if (
					maybeConfig.renderers.some(
						(renderer) => typeof renderer.options.outputDirectory !== "string",
					)
				) {
					errors.push(
						`All ${chalk.cyan(
							"renderers[n].options.outputDirectory",
						)} should be of type ${chalk.cyan("string")}`,
					);
				}
			}
		}

		if (errors.length) {
			throw new Error(
				`Invalid ${chalk.cyan(this.name)} at ${chalk.cyan(
					this.cwd,
				)}:\n\n- ${errors.join("\n- ")}`,
			);
		}
	}
}
