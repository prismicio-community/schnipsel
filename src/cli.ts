import cac from "cac";
import chalk from "chalk";
import exit from "exit";
import latestVersion from "latest-version";
import semver from "semver";

import * as commands from "./commands";
import { NAME, PACKAGE, VERSION, CONFIG_FILE } from "./const";
import { lineBreak, logger } from "./lib";

const cli = cac(NAME);

cli
	.command(
		"",
		`Render Schnipsel project according to current working directory ${chalk.cyan(
			CONFIG_FILE,
		)}`,
		{ allowUnknownOptions: false },
	)
	.option("-w, --watch", "Render in watch mode")
	.action(async (options) => {
		await commands._default(cli, options);
	});

cli
	.command("init", `Init a ${chalk.cyan(CONFIG_FILE)} file`, {
		allowUnknownOptions: false,
	})
	.action(async (options) => {
		await commands.init(cli, options);
	});

cli.version(VERSION);
cli.help(commands.help);

const run = async (): Promise<void> => {
	try {
		cli.parse(process.argv, { run: false });
		await cli.runMatchedCommand();
	} catch (error) {
		if (error instanceof Error) {
			logger.error(error.message);
		} else {
			logger.fatal("");
			console.log(error);
			lineBreak();
		}
		exit(1);
	}

	try {
		const version = await latestVersion(PACKAGE);
		if (semver.gt(version, VERSION)) {
			lineBreak();
			logger.info(
				`A new version of ${chalk.cyan(
					PACKAGE,
				)} is available!\n\n  Install it with:\n  $ npm install --global ${PACKAGE}@${version}\n`,
			);
		}
	} catch (error) {
		// Ignore
	}
};

process.on("unhandledRejection", (error) => {
	logger.fatal("");
	console.log(error);
	lineBreak();
	exit(2);
});

run();
