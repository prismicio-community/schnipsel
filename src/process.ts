import { join } from "node:path";

import chalk from "chalk";
import chokidar from "chokidar";

import { ConfigFile } from "./ConfigFile";
import { Input } from "./Input";
import { Renderers } from "./renderers";
import { logger } from "./lib";
import { Config } from "./types";

export const readAndRender = async (config: Config) => {
	const now = Date.now();

	const input = new Input(config.input);
	const snippets = await input.read();

	logger.info(
		`Rendering ${chalk.cyan(snippets.length)} snippet${
			snippets.length > 1 ? "s" : ""
		} through ${chalk.cyan(config.renderers.length)} renderer${
			config.renderers.length > 1 ? "s" : ""
		}...`,
	);

	await Promise.all(
		config.renderers.map(async (renderer) =>
			new Renderers[renderer.name](renderer.options).render(snippets),
		),
	);

	logger.success(
		`Read and rendered snippets in ${chalk.cyan(`${Date.now() - now}ms`)}`,
	);
};

export const run = async () => {
	const configFile = new ConfigFile();
	const config = await configFile.read();

	await readAndRender(config);
};

export const runWatch = async () => {
	const configFile = new ConfigFile();
	const config = await configFile.read();

	const handler = async () => {
		console.clear();
		const now = new Date();
		logger.info(
			`Changes detected, rendering... (${chalk.cyan(
				`${`00${now.getHours()}`.slice(-2)}:${`00${now.getMinutes()}`.slice(
					-2,
				)}:${`00${now.getSeconds()}`.slice(-2)}`,
			)})`,
		);
		await readAndRender(config);
	};

	chokidar
		.watch("**/*.md", {
			cwd: join(process.cwd(), config.input.directory),
			persistent: true,
		})
		.on("change", handler)
		.on("unlink", handler)
		.on("ready", handler);
};
