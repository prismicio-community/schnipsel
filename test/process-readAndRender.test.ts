import { join } from "node:path";
import { writeFile } from "node:fs/promises";

import { test, expect, vi } from "vitest";

import validConfig from "./__fixtures__/valid/schnipsel.config";

import { readAndRender } from "../src/process";

vi.mock("../src/lib", async () => {
	const {
		logger,
		...rest
	}: Record<string, Record<string, unknown>> = await vi.importActual(
		"../src/lib",
	);

	return {
		logger: { ...logger, warn: vi.fn(), info: vi.fn(), success: vi.fn() },
		...rest,
	};
});

vi.mock("node:fs/promises", async () => {
	const { writeFile: _writeFile, ...rest }: Record<string, unknown> =
		await vi.importActual("node:fs/promises");

	return { writeFile: vi.fn(), ...rest };
});

test("reads and writes snippet files according to config", async () => {
	await readAndRender(validConfig, join(__dirname, "__fixtures__/valid"));

	expect(vi.mocked(writeFile)).toHaveBeenCalledTimes(
		2 /* VS Code */ +
			1 /* Sublime */ +
			1 /* IntelliJ */ +
			2 /* package.json + plugin.xml */,
	);
	expect(
		vi
			.mocked(writeFile)
			// @ts-expect-error - Apparently Vitest is messing types(?)
			.calls.map(([_, content]) => content)
			.sort()
			.join("\n\n\n"),
	).toMatchSnapshot();

	vi.mocked(writeFile).mockClear();
});
