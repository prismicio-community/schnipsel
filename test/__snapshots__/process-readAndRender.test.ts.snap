// Vitest Snapshot v1

exports[`reads and writes snippet files according to config 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<idea-plugin>
	<extensions defaultExtensionNs=\\"com.intellij\\">
		<!--SCHNIPSEL-->
		<defaultLiveTemplates file=\\"/liveTemplates/typescript.xml\\"/>
		<!--ENDSCHNIPSEL-->
	</extensions>
</idea-plugin>



<snippet>
	<content><![CDATA[

]]></content>
	<tabTrigger>helloWorldMissing</tabTrigger>
	<scope>source.ts</scope>
	<description>Says hello to the world missing</description>
</snippet>



<snippet>
	<content><![CDATA[
export const hello = (str: string): void => {
	console.info(\\"Hello %s\\", str);
};
]]></content>
	<tabTrigger>helloWorld</tabTrigger>
	<scope>source.ts</scope>
	<description>Says hello to the world</description>
</snippet>



<templateSet group=\\"typescript\\">
	<template
		name=\\"helloWorld\\"
		value=\\"export const hello = (str: string): void =&gt; {&#10;	console.info(&quot;Hello %s&quot;, str);&#10;};\\"
		description=\\"Says hello to the world\\"
		toReformat=\\"true\\"
		toShortenFQNames=\\"true\\"
	>
		<context>
			<option name=\\"TYPESCRIPT\\" value=\\"true\\"/>
		</context>
	</template>
	<template
		name=\\"helloWorldMissing\\"
		value=\\"\\"
		description=\\"Says hello to the world missing\\"
		toReformat=\\"true\\"
		toShortenFQNames=\\"true\\"
	>
		<context>
			<option name=\\"TYPESCRIPT\\" value=\\"true\\"/>
		</context>
	</template>
</templateSet>



{
  \\"Hello World\\": {
    \\"scope\\": \\"typescript\\",
    \\"prefix\\": \\"helloWorld\\",
    \\"description\\": \\"Says hello to the world\\",
    \\"body\\": [
      \\"export const hello = (str: string): void => {\\",
      \\"\\\\tconsole.info(\\\\\\"Hello %s\\\\\\", str);\\",
      \\"};\\"
    ]
  },
  \\"Hello World Missing\\": {
    \\"scope\\": \\"typescript\\",
    \\"prefix\\": \\"helloWorldMissing\\",
    \\"description\\": \\"Says hello to the world missing\\",
    \\"body\\": [
      \\"\\"
    ]
  }
}



{
  \\"contributes\\": {
    \\"snippets\\": [
      {
        \\"language\\": \\"typescript\\",
        \\"path\\": \\"./snippets/typescript.code-snippets\\"
      }
    ]
  }
}
"
`;
